numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -985,6 +985,54 @@ PyArray_CanCastArrayTo(PyArrayObject *arr, PyArray_Descr *to,
     return PyArray_CanCastTypeTo(from, to, casting);
 }
 
+
+NPY_NO_EXPORT const char *
+npy_casting_to_string(NPY_CASTING casting)
+{
+    switch (casting) {
+        case NPY_NO_CASTING:
+            return "'no'";
+        case NPY_EQUIV_CASTING:
+            return "'equiv'";
+        case NPY_SAFE_CASTING:
+            return "'safe'";
+        case NPY_SAME_KIND_CASTING:
+            return "'same_kind'";
+        case NPY_UNSAFE_CASTING:
+            return "'unsafe'";
+        default:
+            return "<unknown>";
+    }
+}
+
+
+/**
+ * Helper function to set a useful error when casting is not possible.
+ *
+ * @param src_dtype
+ * @param dst_dtype
+ * @param casting
+ * @param scalar Whether this was a "scalar" cast (includes 0-D array with
+ *               PyArray_CanCastArrayTo result).
+ */
+NPY_NO_EXPORT void
+npy_set_invalid_cast_error(
```

Comment:
I wonder if this function is even worth it at this point, given it would be a one-liner at each call site.

Reply:
Yeah, have to admit, I am not sure. On the other hand, that way at least the error message is not spelled out in 5 places.

Reply:
Missed how many times this was used, that seems fair. Maybe replace `scalar` with `char *what`, which makes it possible for the caller to pass more info to put in the message, which will also make the implementation shorter.

Reply:
To be fair, 2 of those would be the concatenate fixup...

Reply:
Made that simplifictaion for the iterator code (which is arguably very similar), and adds another 1.5 occurances, which then are not covered here (so we have 2 "places" in the code). 

The only real reason for this being useful is probably the scalar distinction, which is correct, but a bit wonky admittedly.